Add pool DB operation `listRetiredPools`.

jonathanknowles opened the pull request:
# Issue Number

#2016 

# Overview

This PR:

- [x] Adds a `listRetiredPools` operation to the pool DB layer:<br><br>
    ```hs
    listRetiredPools
        :: EpochNo
        -> stm [PoolId]
        -- ^ List all pools with an active retirement epoch that is
        -- earlier than or equal to the specified epoch.
    ````
- [x] Provides the following implementations of `listRetiredPools`.
    - [x] `Pool.DB.MVar`
    - [x] `Pool.DB.SQLite`.

The SQLite implementation is based on an **SQL view** ([see definition](https://github.com/input-output-hk/cardano-wallet/pull/2024/files#diff-08afa2852f50fa31f5757942d004ce8aR513)) which determines the currently-active set of retirement certificates.

<details><summary>Click to view sample data</summary>

![active_pool_retirements](https://user-images.githubusercontent.com/206319/90096831-13b1a280-dd67-11ea-9612-fd1c4a825fb6.png)

</details>

# Property Testing

This PR also adds property tests to check that:
- [x] `listRetiredPools` **does not** return:
    - [x] pools that have never had retirement certificates associated with them.
    - [x] pools that have had their most-recently-published retirement certificates revoked by subsequent re-registration certificates.
    - [x] pools with effective retirement epochs that are later than the specified epoch.
- [x] `listRetiredPools` returns the correct set of pools in the following contexts:
    - [x] where there are **multiple pools**.
    - [x] where there are **multiple certificates per pool**
        (registrations, retirements, re-registrations, re-retirements).
    - [x] where the publications of certificates affecting different pools are **interleaved together in time**.

# Mutation Testing

To increase confidence that the `activePoolRetirements` view is correctly defined, I have manually performed the following mutations and verified that the `prop_listRetiredPools_multiplePools_multipleCerts` property fails appropriately.

### Mutation 1

```patch
-    WINDOW w AS (ORDER BY pool_id, slot desc, slot_internal_index desc)
+    WINDOW w AS (ORDER BY pool_id, slot     , slot_internal_index desc)
 )
 GROUP BY pool_id
```

### Mutation 2

```patch
-    WINDOW w AS (ORDER BY pool_id, slot desc, slot_internal_index desc)
+    WINDOW w AS (ORDER BY pool_id, slot desc, slot_internal_index     )
 )
 GROUP BY pool_id
```

### Mutation 3

```patch
-    WINDOW w AS (ORDER BY pool_id, slot desc, slot_internal_index desc)
+    WINDOW w AS (ORDER BY pool_id, slot     , slot_internal_index     )
 )
 GROUP BY pool_id
```

### Mutation 4

```patch
-    WINDOW w AS (ORDER BY pool_id, slot desc, slot_internal_index desc)
+    WINDOW w AS (ORDER BY pool_id, slot desc,                         )
 )
 GROUP BY pool_id
```

jonathanknowles committed:
Add dependency on `string-qq` package.

jonathanknowles committed:
Add pool database view `active_pool_retirements`.

jonathanknowles committed:
Add skeleton `listRetiredPools` pool DB operation.

jonathanknowles committed:
Provide model implementation of `listRetiredPools` operation.

jonathanknowles committed:
Provide SQLite implementation of `listRetiredPools` operation.

jonathanknowles committed:
Clarify description of `prop_readPoolLifeCycleStatus`.

jonathanknowles committed:
Add accessor functions {get,set}PoolCertificatePoolId.

jonathanknowles committed:
Add type `SinglePoolCertificateSequence`.

jonathanknowles committed:
Add coverage property test for `SinglePoolCertificateSequence`.

jonathanknowles committed:
Use `SinglePoolCertificateSequence` in `prop_readPoolLifeCycleStatus`.

jonathanknowles committed:
Use `assertWith` in `prop_readPoolLifeCycleStatus`.

jonathanknowles committed:
Extend `prop_readPoolLifeCycleStatus`.

Extend `prop_readPoolLifeCycleStatus` to call `listRetiredPools`.

jonathanknowles committed:
Simplify property `prop_readPoolLifeCycleStatus`.

jonathanknowles committed:
Use `assertWith` in `prop_multiple_putPoolRetirement_single_readPoolRetirement`.

jonathanknowles committed:
Extend `prop_multiple_putPoolRetirement_single_readPoolRetirement`.

Extend `prop_multiple_putPoolRetirement_single_readPoolRetirement` to
call `listRetiredPools`.

jonathanknowles committed:
Use `assertWith` in `prop_multiple_putPoolRegistration_single_readPoolRegistration`.

jonathanknowles committed:
Extend `prop_multiple_putPoolRegistration_single_readPoolRegistration`.

Extend `prop_multiple_putPoolRegistration_single_readPoolRegistration`
to call `listRetiredPools`.

jonathanknowles committed:
Use `assertWith` in `prop_pool{Registration,Retirement}`.

jonathanknowles committed:
Extend `prop_pool{Registration,Retirement}`.

Extend `prop_pool{Registration,Retirement}` to call `listRetiredPools`.

jonathanknowles committed:
Use `assertWith` in `prop_rollbackRetirement`.

jonathanknowles committed:
Extend `prop_rollbackRetirement`.

Extend `prop_rollbackRetirement` to call `listRetiredPools`.

jonathanknowles committed:
Add test utility function `interleave`.

KtorZ commented on lib/core/src/Cardano/Pool/DB/Sqlite.hs:
```diff
@@ -347,6 +355,20 @@ newDBLayer trace fp timeInterpreter = do
+            let safeCast (Single poolId, Single retirementEpoch) =
+                    PoolRetirementCertificate
+                        <$> fromPersistValue poolId
+                        <*> fromPersistValue retirementEpoch
```
Hmmm. Why is the unsafe query needed here? We should stick with the persistent DSL as much as possible and only resort to unsafe queries when the DSL is too limiting or becomes a real performance impediment.

jonathanknowles replied:
@KtorZ wrote:
> Why is the unsafe query needed here?

When determining the current set of effective retirements, we need to combine results from **two different tables**:

1. `pool_registrations`
2. `pool_retirements`

This is because registration and retirement certificates can _cancel each other out_.

The only way to find out whether a retirement certificate is still in effect is to verify that there are no other retirement certificates _or_ registration certificates that cancel it out.

> we should only resort to unsafe queries when the DSL is too limiting

I do agree with this.

However, the implementation in this PR determines the set of currently-active retirement certificates with a **single query** that uses:
- the SQL `UNION` function;
- the SQL `row_number` window function.

As far as I know, `persistent` supports neither of the above functions (at least out of the box).

Furthermore, defining this query in terms of an SQL **view** can help us if we need to debug a real pool database. By opening the database and fetching the `active_pool_retirements` view, we can immediately see which retirement certificates the wallet currently considers to be active.

jonathanknowles replied:
Here's the `active_pool_retirements` view for mainnet:

![active_pool_retirements](https://user-images.githubusercontent.com/206319/90096811-01cfff80-dd67-11ea-8ada-920d24f10e20.png)

To implement `listRetiredPools`, we just need to select all rows from this view with a `retirement_epoch` that is _earlier than or equal_ to the given epoch argument.

jonathanknowles committed:
Add property `prop_listRetiredPools_multiplePools_multipleCerts`.

jonathanknowles committed:
Use different certificate serialization orders.

Use different serialization orders when serializing certificate sequences from
multiple pools in property `prop_listRetiredPools_multiplePools_multipleCerts`.

This gives us extra confidence that the `activeRetirementCertificates` view is
handling the `(slot, slot_internal_index)` order correctly.

KtorZ commented on lib/core/test/unit/Cardano/Pool/DB/Properties.hs:
```diff
@@ -465,13 +477,20 @@ prop_poolRegistration DBLayer {..} entries =
+        assertWith "entriesIn == entriesOut"
+            $ entriesIn == entriesOut
+        assertWith "no pools are marked to retire"
+            $ null poolsMarkedToRetire
```
:+1:

KtorZ commented on lib/core/test/unit/Cardano/Pool/DB/Properties.hs:
```diff
@@ -488,13 +507,22 @@ prop_poolRetirement DBLayer {..} entries =
+        assertWith "all pools are marked to retire"
+            $ (==)
+                (Set.fromList $ snd <$> entriesIn)
+                (Set.fromList poolsMarkedToRetire)
```
:+1:

KtorZ commented on lib/core/test/unit/Cardano/Pool/DB/Properties.hs:
```diff
@@ -526,9 +556,11 @@ prop_multiple_putPoolRegistration_single_readPoolRegistration
             mRetrievedCertificatePublication
             mExpectedCertificatePublication
+        assertWith "pool is not marked to retire" $
+            null poolsMarkedToRetire
```
:+1:

KtorZ commented on lib/core/test/unit/Cardano/Pool/DB/Properties.hs:
```diff
@@ -822,6 +871,91 @@ prop_listRegisteredPools DBLayer {..} entries =
+        lifeCycleStatuses <- run $ atomically $ do
+            mapM readPoolLifeCycleStatus allPoolIds
+        let poolsMarkedToRetire = catMaybes $
+                getPoolRetirementCertificate <$> lifeCycleStatuses
```
I'd be better not to rely on `lifeCycleStatuses` to determine which pools are supposed to be retired. The risk is to have a similar logic bug in both `lifeCycleStatuses` and `listRetiredPools` that remain uncaught, especially if the same approach is used behind the scene.

jonathanknowles replied:
> I'd be better not to rely on `lifeCycleStatuses` to determine which pools are supposed to be retired. The risk is to have a similar logic bug in both `lifeCycleStatuses` and `listRetiredPools` that remain uncaught, especially if the same approach is used behind the scene.

I definitely agree, this is indeed a risk.

To counteract this risk:

1. We do have a suite of tests specifically aimed at `determinePoolLifeCycleStatus`:
    - `prop_determinePoolLifeCycleStatus_orderCorrect`
    - `prop_determinePoolLifeCycleStatus_neverRegistered`
    - `prop_determinePoolLifeCycleStatus_differentPools`<br><br>

2. The definition of `listRetiredPools` uses a completely different approach in the form of a handwritten query. In the property test for this query, we check it against the `determinePoolLifeCycleStatus` function, which we consider to be the source of truth.
